### DISCUSSION TITLE
Use Bunny Fonts instead of Google Fonts

### DISCUSSION THREAD

**Discussion by jonaharagon** [squidfunk/mkdocs-material#5475] [Ideas]
Google Fonts is not GDPR compliant. [Bunny Fonts](https://fonts.bunny.net/about) is a GDPR compliant drop-in replacement for Google Fonts. Bunny Fonts is fully compatible with the Google Fonts CSS v1 API, meaning the only change required would be swapping `https://fonts.googleapis.com/css` for `https://fonts.bunny.net/css`.

I would like to see this change, because it would give website operators the option to use a globally available font provider without having to subject their users to Google's tracking, which is in line with the "fast and lightweight" project philosophy.

**Comment by squidfunk** [comment]
Thanks for suggesting. This is a pretty fundamental change, as we would be switching to an entirely different CDN. Checking their font explorer, Bunny Fonts does not seem to be a full replacement, as Google Fonts has 1,508 and Bunny Fonts has 1,492 fonts. I'm not sure how they sync fonts and functionality, but it doesn't seem to be a transparent proxy.

IMHO, swapping out Google Fonts for another solution is only solving part of the problem. The canonical way to achieve full GDPR compliance is to use the [built-in privacy plugin](https://squidfunk.github.io/mkdocs-material/setup/ensuring-data-privacy/#built-in-privacy-plugin), which will automatically download and inline all external assets. However, since you would only need to replace an URL, you could also just use a [hook](https://www.mkdocs.org/user-guide/configuration/#hooks) and replace the URL, and be done:

``` python
def on_post_page(output, page, config):
    return output.replace(
        "https://fonts.googleapis.com/css",
        "https://fonts.bunny.net/css"
    )
```

Then save that to `hooks/fonts.py` and add the following to `mkdocs.yml`:

``` yaml
hooks:
  - hooks/fonts.py
```

Four lines of code, so it's ridiculously simple.

**Reply by kamilkrzyskow** [reply]
Aren't the fonts stored in the page head tag? If so instead of using `on_page_content` I think `on_post_page` is required 🤔
Also the str.replace isn't a in place operation. I'm on my phone sorry for incomplete formatting:
```py
html = html.replace(old, new, amount)
return html
```
Store the replaced value, and return it from the function to let MkDocs save it.

Also to increase performance of processing every file consider setting an upper maximum amount of replacements.

Also for anyone interested I have a guide [how to dynamically modify templates](https://github.com/squidfunk/mkdocs-material/discussions/5060). Using my guide to change the values in the base.html would allow to greatly limit the amount of processing.
However, use the newer version from the repository.

I still have to fix the code redundancy and update the guide 😓

**Reply by squidfunk** [reply]
Thanks, fixed my example!

**Reply by kamilkrzyskow** [reply]
Sure, np. However, while changing the signature you kept the on_page_content name.

**Reply by squidfunk** [reply]
Fixed again. Sorry, busy day with the social cards release. Doing too many things at once 😅
